src/material_library.h

```diff
@@ -0,0 +1,207 @@
+  /**
+   * \brief determines in that datapath exists in the hdf5 file
+   * \param[in] filename of the h5m file
+   * \param[in] the datapath of we would like to test
+   * \return true/false
+   */
+  bool hdf5_path_exists(const std::string& filename,
+                        const std::string& datapath);
+
+  /**
+   * \brief determines the length of an hdf5 data table
+   * \param[in] filename of the h5m file
+   * \param[in] the datapath of we would like to read
+   * \return the number of elements to the array
+   */
+  int get_length_of_table(const std::string& filename,
+                          const std::string& datapath);
+
+ public:
+  // materialLibrary constructor
+  MaterialLibrary();  //< empty constryctor
+
+  /**
+   * Constructor from file
+   * \param filename path to file on disk, this file may be either in plaintext
+   *                 or HDF5 format.
+   * \param datapath Path to the materials in the file.
+   * \param nucpath Path to the nuclides set in the file.
+  */
+  MaterialLibrary(const std::string& filename,
+                  const std::string& datapath = "/materials",
+                  const std::string& nucpath = "");
+
+  ~MaterialLibrary() {};  //< default destructor
+
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the HDF5 file.
+    /// \param datapath Path to the materials in the file.
+    /// \param nucpath Path to the nuclides set in the file.
+    /// \param protocol Flag for layout of material on disk.
+  */
+  void from_hdf5(const std::string& filename,
+                 const std::string& datapath = "/materials",
+                 const std::string& nucpath = "", int protocol = 1);
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the HDF5 file.
+    /// \param datapath Path to the materials in the file.
+    /// \param nucpath Path to the nuclides set in the file.
+    /// \param protocol Flag for layout of material on disk.
+  */
+  void from_hdf5(char* filename, char* datapath = "/materials" , char* nucpath = "", int protocol = 1);
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the json file.
+  */
+  void from_json(char* filename);
+  /**
+   * \brief loads the pyne materials in map of name vs Material
+    /// \param filename Path on disk to the json file.
+  */
+  void from_json(const std::string& filename);
+  void load_json(Json::Value json);
+  Json::Value dump_json();
+  void write_json(const std::string& filename);
+  void write_json(char* filename);
+  /**
+   * Writes MaterialLibrary out to an HDF5 file.
+   *  This happens according to protocol 1.
+   *  Writting in a file already containing 
+   *  the nuclist (same name) will be overwritten 
+   *  the existing one. This might cause issue to 
+   *  read material already present in the datapath, 
```

Comment:
This is a method to write, so it is not clear why reading matters.

Reply:
my guess is in some cases, if you overwrite the nucpath, then the pre-existing material in the data path, might get messed up.

and I am not quite sure about the results of reading back the datapath

Reply:
First, this comment should then describe corruption of the data.  It's not that there is a problem reading it, it's that the data will be unreliable after the write.

Second, if this happens, it's a problem.  We need to ensure that there is no data corruption.

Reply:
Basically, we should completely delete and replace the material library in an HDF5 every time that we write it.  All merging/adding/deleting of materials should happen in memory and not at the time of writing.